Conflict-resolution-perf-improvements.md

```diff
@@ -0,0 +1,128 @@
+Conflict Resolution Performance Improvements
+====================
+Currently conflict resolution (CR) performance is dog slow, due to the following:
+
+1. All segments and their cache entities are checked
+    - Performance degrades rapidly as the size of the cluster increases
+
+2. Centralised CR
+    - The merge coordinator is responsible for comparing all cache entities
+        - In a distributed cache, this requires the coordinator to retrieve all segments from both partitions in order
+        to compare them
+    - User triggered CR occurs on the node the request is made
+
+3. No parallelism
+    - A single segment is processed at a given time in order to avoid the centralised CR coordinator suffering a OOM exception
+
+# Proposals
+The following proposals are listed in the order of the anticipated performance benefit, with those with greatest impact
+presented first.
+
+## Maintaining a Merkle tree
+To minimise the number of entries that need to be checked during CR, we should maintain a Merkle tree of cache entry hashes
+on a per segment basis, with the root of the tree being a hash of all the entry hashes in the segment. Given the root hash
+of a segment from three different nodes, we can determine that no conflict exists simply by checking if all three values
+are equal. If any of the three hashes differ, then it's then necessary to compare the segment's entries and perform CR.
+Amazon Dynamo DB utilises this technique https://www.allthingsdistributed.com/files/amazon-dynamo-sosp2007.pdf.
+
+The simplest solution for producing such a tree, is to simple have a tree of depth 2. Where each cache entry is a leaf node
+and the root is a hash of all leaf hash values. This can be easily implemented by simply iterating over the segment container
+in order to calculate the hash of individual entries and the combined hash.
+
+TRACKED BY: [ISPN-8412] https://issues.redhat.com/browse/ISPN-8412
```

Comment:
Wrong link syntax, should be `[ISPN-8412](https://issues.redhat.com/browse/ISPN-8412)`